openstack/identity/v3/extensions/ec2tokens/requests.go

```diff
@@ -0,0 +1,279 @@
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/tokens"
+)
+
+const (
+	aws4Request  = "aws4_request"
+	v2AlgoSha1   = "HmacSHA1"
+	v2AlgoSha256 = "HmacSHA256"
+	v4Algo       = "AWS4-HMAC-SHA256"
+	iso8601utc   = "20060102T150405Z"
+	rfc3339utc   = "2006-01-02T15:04:05Z"
+	yyyymmdd     = "20060102"
+)
+
+// AuthOptions represents options for authenticating a user using EC2 credentials.
+type AuthOptions struct {
+	// EC2 access ID.
+	Access string `json:"access" required:"true"`
+	// EC2 access secret, used to calculate signature only.
+	// Not used, when a Signature is provided.
+	Secret string `json:"-"`
+	// Optional parameters.
+	Host    string            `json:"host"`
+	Path    string            `json:"path"`
+	Verb    string            `json:"verb"`
+	Headers map[string]string `json:"headers"`
+	Region  string            `json:"-"`
+	Service string            `json:"-"`
+	Params  map[string]string `json:"params"`
+	// Allows Gophercloud to re-authenticate automatically if/when your
+	// token expires.
+	AllowReauth bool `json:"-"`
+	// Can be either a []byte (encoded to base64 automatically) or a
+	// string. You can set the singature explicitly, when you already know
+	// it.
+	Signature interface{} `json:"signature"`
+	// Optional random hash, when signature is nil.
+	BodyHash *string `json:"body_hash"`
+	// Optional timestamp to calculate a V4 signature.
+	Timestamp *time.Time `json:"-"`
+}
+
+func canonicalQsV2(params map[string]string) string {
+	var keys []string
+	for k := range params {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	var pairs []string
+	for _, k := range keys {
+		pairs = append(pairs, fmt.Sprintf("%s=%s", k, url.QueryEscape(params[k])))
+	}
+
+	return strings.Join(pairs, "&")
+}
+
+func stringToSignV2(opts AuthOptions, params map[string]string) string {
+	stringToSign := strings.Join([]string{
+		opts.Verb,
+		opts.Host,
+		opts.Path,
+	}, "\n")
+
+	return strings.Join([]string{
+		stringToSign,
+		canonicalQsV2(params),
+	}, "\n")
+}
+
+func canonicalQsV4(verb string, params map[string]string) string {
+	if verb == "POST" {
+		return ""
+	}
+	return canonicalQsV2(params)
+}
+
+func canonicalHeadersV4(headers map[string]string, signedHeaders string) string {
+	headersLower := make(map[string]string, len(headers))
+	for k, v := range headers {
+		headersLower[strings.ToLower(k)] = v
+	}
+
+	var headersList []string
+	for _, h := range strings.Split(signedHeaders, ";") {
+		if v, ok := headersLower[h]; ok {
+			headersList = append(headersList, h+":"+v)
+		}
+	}
+
+	return strings.Join(headersList, "\n") + "\n"
+}
+
+func sumHMAC1(key []byte, data []byte) []byte {
+	hash := hmac.New(sha1.New, key)
+	hash.Write(data)
+	return hash.Sum(nil)
+}
+
+func sumHMAC256(key []byte, data []byte) []byte {
+	hash := hmac.New(sha256.New, key)
+	hash.Write(data)
+	return hash.Sum(nil)
+}
+
+func signatureKeyV4(secret, date, region, service string) []byte {
```

Comment:
We should name this in an exportable and testable way along with including a doc string and unit test:

```
// EC2CredentialsBuildSignatureKeyV4 builds...
// https://github.com/openstack/python-keystoneclient/blob/526350eaeeadef8c1ca5f80547c01f94342a6653/keystoneclient/contrib/ec2/utils.py#L169
```

Reply:
done